databuilder/query_engines/base_sql.py

```diff
@@ -76,187 +128,252 @@ class BaseSQLQueryEngine(BaseQueryEngine):
+            # Ensure the results_query JOINs on all the tables it needs to be able to
+            # include this column in the output
+            results_query = include_joined_tables(
+                results_query,
+                get_referenced_tables(column_query),
+                join_column="patient_id",
+            )
+            # Add this column to the final selected results using the supplied name
+            results_query = results_query.add_columns(column_query.label(column_name))
 
-        # Generate the big query that creates the base population table and its columns,
-        # selected from the output group tables
-        results_query = self.generate_results_query()
+        # Get all the setup queries needed to populate the various temporary tables used
+        # by the results_query, and the queries needed to clean them up afterwards
+        setup_queries, cleanup_queries = get_setup_and_cleanup_queries(results_query)
 
-        temp_table_queries = sum(self.temp_tables.values(), start=[])
-        return temp_table_queries + [results_query]
+        return setup_queries, results_query, cleanup_queries
 
     @contextlib.contextmanager
     def execute_query(self):
-        queries = self.get_queries()
+        setup_queries, results_query, cleanup_queries = self.get_queries()
         with self.engine.connect() as cursor:
-            for query in queries:
-                result = cursor.execute(query)
-            # We're only interested in the results from the final query
-            yield result
-            self.post_execute_cleanup(cursor)
+            for query in setup_queries:
+                cursor.execute(query)
 
-    def post_execute_cleanup(self, cursor):
-        """
-        Called after results have been fetched
-        """
-        for table in self.temp_tables.keys():
-            self.drop_temp_table(cursor, table)
+            yield cursor.execute(results_query)
+
+            for query in cleanup_queries:
+                cursor.execute(query)
 
-    def drop_temp_table(self, cursor, table):
+    def get_sql_element(self, node: QueryNode) -> ClauseElement:
         """
-        Drop the specified temporary table
+        Caching wrapper around `get_sql_element_no_cache()` below, which is the
+        entrypoint method for converting QueryNodes to SQL ClauseElements
+
+        The caching is a peformance enhancement, not for the Python code, but rather for
+        the SQL we generate. It ensures that we don't get needlessly complex (though
+        correct) SQL by generating duplicated temporary tables or table-like clauses.
         """
-        query = sqlalchemy.schema.DropTable(table, if_exists=True)
-        cursor.execute(query)
+        if node in self.sql_element_cache:
+            return self.sql_element_cache[node]
+        else:
+            element = self.get_sql_element_no_cache(node)
+            self.sql_element_cache[node] = element
+            return element
```

Comment:
That was my original thought but there were a couple of reasons I didn't. One is that because of the `singledispatch` stuff we can't apply it directly to the method we want to cache. So we'd still need a separate method like:
```py
@cache
def get_sql_element(self, node):
    return self.get_sql_element_no_cache(node)
```
At which point the neatness of doing it via a decorator is lost a bit.

The other is that `functools.cache` actually creates a single shared cache across instances of that class (and all subclasses). But because `self` is included as a cache key it ends up behaving like a per-instance cache. Mostly this is fine, but it has some unexpected consequences like references hanging around forever even after the instance is gone, and `method.clear_cache()` affecting all instances. None of that particularly matters here, but it makes me feel icky about using it.